### PR TITLE
Bug 1810776 - Add SwipeToDismiss to composed tabs tray

### PR DESCRIPTION
**Pull request by Alexandru2909**
Added swipe to dismiss gesture to composed tabs tray.

| Current xml behavior | Composed behavior |
| --- | --- |
| <video src="https://github.com/mozilla-mobile/firefox-android/assets/35462038/b34fbcc8-e5c2-4adb-882d-f9e6ddf4932b" /> | <video src="https://github.com/mozilla-mobile/firefox-android/assets/35462038/7c5968de-0a49-4114-8041-ea0093eb14ca" /> |

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1810776

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            modifier = Modifier.matchParentSize(),
+        )
+        Row(
+            content = dismissContent,
+            modifier = Modifier.offset { IntOffset(state.offset.value.roundToInt(), 0) },
+        )
+    }
+}
+
+private const val DISMISS_THRESHOLD = 0.5f
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt:93]: nit: could we get a quick demo preview here?

File: fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * A composable that can be dismissed by swiping left or right
+ *
+ * @param state The state of this component.
+ * @param modifier Optional [Modifier] for this component.
+ * @param enabled [Boolean] controlling whether the content is swipeable or not.
+ * @param directions The set of directions in which the component can be dismissed.
+ * @param dismissThreshold The threshold the item needs to be swiped in order to be dismissed.
+ * @param background A composable that is stacked behind the content and is exposed when the
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt:42]:
```suggestion
 * @param backgroundContent A composable that is stacked behind the primary content and is exposed when the
```

File: fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                enabled = state.currentValue == Default,
+                reverseDirection = LocalLayoutDirection.current == LayoutDirection.Rtl,
+                resistance = null,
+            )
+            .then(modifier),
+    ) {
+        Row(
+            content = background,
+            modifier = Modifier.matchParentSize(),
+        )
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt:85]: nit: newline between to the two `Row`s

File: fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        LocalConfiguration.current.screenWidthDp.dp.toPx()
+    }
+
+    val anchors = mutableMapOf(0f to Default)
+    if (StartToEnd in directions) anchors += swipeWidth to DismissedToEnd
+    if (EndToStart in directions) anchors += -swipeWidth to DismissedToStart
+
+    val thresholds = { _: DismissValue, _: DismissValue ->
+        dismissThreshold
+    }
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt:67]: nit: group the vals together then do the anchor maths
```suggestion
    val swipeWidth = with(LocalDensity.current) {
        LocalConfiguration.current.screenWidthDp.dp.toPx()
    }
    val anchors = mutableMapOf(0f to Default)
    val thresholds = { _: DismissValue, _: DismissValue ->
        dismissThreshold
    }
    
    if (StartToEnd in directions) anchors += swipeWidth to DismissedToEnd
    if (EndToStart in directions) anchors += -swipeWidth to DismissedToStart
```

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/DismissedTabBackground.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import androidx.compose.ui.graphics.Shape
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.unit.dp
+import mozilla.components.feature.tab.collections.Tab
+import org.mozilla.fenix.R
+import org.mozilla.fenix.compose.annotation.LightDarkPreview
+import org.mozilla.fenix.theme.FirefoxTheme
+
+/**
+ * Composable used to display the background of a [Tab] shown in collections that is being swiped left or right.
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/DismissedTabBackground.kt:32]:
```suggestion
 * Thebackground of a [Tab] that is being swiped left or right.
```

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/DismissedTabBackground.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * @param dismissDirection [DismissDirection] of the tab being swiped depending on which this composable
+ * will also indicate the swipe direction by placing a warning icon at the start of the swipe gesture.
+ * If `null` the warning icon will be shown at both ends.
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/DismissedTabBackground.kt:36]: I know you're just moving this Composable to its own file, but could you double check the wording of the kdoc for `dismissDirection`? I think I'm misunderstanding what this parameter is driving.

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt [hunk 7]
```diff
@@ -71,7 +76,7 @@ import org.mozilla.fenix.theme.FirefoxTheme
  * @param onClick Callback to handle when item is clicked.
  * @param onLongClick Callback to handle when item is long clicked.
  */
-@OptIn(ExperimentalFoundationApi::class)
+@OptIn(ExperimentalFoundationApi::class, ExperimentalMaterialApi::class)
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt:79]: which line needed `ExperimentalMaterialApi` to be added?

**Review by MozillaNoah — COMMENTED** [review]
first half-pass of some nits before I break for lunch. I haven't had a chance to pull this down yet

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt (hunk 7 above)
Comment (Alexandru2909) [reply fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt:79]: `ExperimentalMaterialApi` is required for `DismissState`

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/DismissedTabBackground.kt [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                    ),
+                tint = FirefoxTheme.colors.iconWarning,
+            )
+        }
+    }
+}
+
+@Composable
+@LightDarkPreview
+private fun TabInCollectionPreview() {
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/DismissedTabBackground.kt:83]:
```suggestion
private fun DismissedTabBackgroundPreview() {
```

File: fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+private const val DISMISS_THRESHOLD = 0.5f
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+private fun SwipeablePreview(directions: Set<DismissDirection>, text: String) {
+    val state = rememberDismissState()
+
+    FirefoxTheme {
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt:112]: nit: the firefox theme call could live inside `SwipeToDismissPreview` instead.

File: fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+@Composable
+@Preview
+private fun SwipeToDismissPreview() {
+    Column {
+        SwipeablePreview(directions = setOf(StartToEnd), text = "Swipe to right ->")
+        Spacer(Modifier.height(30.dp))
+        SwipeablePreview(directions = setOf(EndToStart), text = "<- Swipe to left")
+        Spacer(Modifier.height(30.dp))
+        SwipeablePreview(directions = setOf(StartToEnd, EndToStart), text = "<- Swipe both ways ->")
+    }
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/SwipeToDismiss.kt:152]: love how succinct this preview is to show how to use SwipeToDismiss! `dismissThreshold` is the only non-obvious parameter not captured you might want to add an example for.

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/DismissedTabBackground.kt [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            Box(modifier = Modifier.height(56.dp)) {
+                DismissedTabBackground(
+                    dismissDirection = DismissDirection.StartToEnd,
+                    shape = RoundedCornerShape(0.dp),
+                )
+            }
+
+            Spacer(Modifier.height(10.dp))
+
+            Box(modifier = Modifier.height(56.dp)) {
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/DismissedTabBackground.kt:95]: can we get one more use case for when both icons are shown?

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt [hunk 12]
```diff
@@ -94,95 +99,115 @@ fun TabGridItem(
...
-                                .clickable { onCloseClick(tab) }
-                                .size(24.dp)
                                 .align(Alignment.CenterVertically)
-                                .testTag(TabsTrayTestTag.tabItemClose),
+                                .padding(start = 8.dp),
                         )
+
+                        HorizontalFadingEdgeBox(
+                            modifier = Modifier
+                                .fillMaxWidth()
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt:159]: was this intentionally changed from `.weight(1f)` to `.fillMaxWidth()`? The close icon isn't showing anymore (the preview captures this visual bug), which is likely also causing the UI test failure.

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt [hunk 13]
```diff
@@ -94,95 +99,115 @@ fun TabGridItem(
...
+            } else {
+                false
+            }
+        },
+    )
+
+    SwipeToDismiss(
+        state = dismissState,
+        enabled = !multiSelectionEnabled,
+        backgroundContent = {},
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt:116]: if you foresee this being unused in more than one location long term, we could make `backgroundContent` optional with `{}` as the default value.

**Review by MozillaNoah — COMMENTED** [review]
this is looking great and it works locally! I left some comments, but I think we may need to add a swipe tolerance to the tab items, as they don't match the XML's swipe behavior. Small, micro swipes will close the tabs in the rewrite vs the XML keeping the tab open.

File: fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt (hunk 13 above)
Comment (Alexandru2909) [reply fenix/app/src/main/java/org/mozilla/fenix/compose/tabstray/TabGridItem.kt:116]: I don't think having the background content as empty will be a common use case, let's keep it as non-optional.

**Comment by Alexandru2909** [comment]
> I think we may need to add a swipe tolerance to the tab items, as they don't match the XML's swipe behavior. Small, micro swipes will close the tabs in the rewrite vs the XML keeping the tab open.

I can't reproduce the issue with small swipes keeping the tab open in the XML version and dismissing it in the rewrite. They behave the same on all my devices, even when using a stylus pen. Could you maybe post a small video with the gesture?

**Comment by MozillaNoah** [comment]
let me know if the UI tests keep failing after your next push. either we can pair on it, or I can push a fix from my end.

**Review by MozillaNoah — APPROVED** [review]

**Pull request merged**